tests/test_single_document.py

```diff
@@ -58,13 +58,28 @@ def test_did_close(self) -> 'Generator':
 
     def test_did_change(self) -> 'Generator':
         assert self.view
+        self.maxDiff = None
         self.insert_characters("A")
         yield from self.await_message("textDocument/didChange")
         # multiple changes are batched into one didChange notification
-        self.insert_characters("B")
-        self.insert_characters("C")
+        self.insert_characters("B\n")
+        self.insert_characters("🙂\n")
         self.insert_characters("D")
-        yield from self.await_message(("textDocument/didChange"))
+        promise = YieldPromise()
+        yield from self.await_message("textDocument/didChange", promise)
+        self.assertEqual(promise.result(), {
+            'contentChanges': [
+                {'range': {'start': {'line': 0, 'character': 1}, 'end': {'line': 0, 'character': 1}}, 'text': 'B'},
+                {'range': {'start': {'line': 0, 'character': 2}, 'end': {'line': 0, 'character': 2}}, 'text': '\n'},
+                {'range': {'start': {'line': 1, 'character': 0}, 'end': {'line': 1, 'character': 0}}, 'text': '🙂'},
+                # Note that this is character offset (2) is correct (UTF-16).
+                {'range': {'start': {'line': 1, 'character': 2}, 'end': {'line': 1, 'character': 2}}, 'text': '\n'},
+                {'range': {'start': {'line': 2, 'character': 0}, 'end': {'line': 2, 'character': 0}}, 'text': 'D'}],
+            'textDocument': {
+                'version': self.view.change_count(),
+                'uri': filename_to_uri(TEST_FILE_PATH)
+            }
+        })
```

Comment:
This is an important test: it verifies that the character offset for the second newline char is **two**, not **one**.